system-x/services/kudu/src/main/java/software/tnb/kudu/resource/openshift/RemoteClient.java

```diff
@@ -0,0 +1,129 @@
+                        throw new RuntimeException("unable to read leader master");
+                    }
+                    final List<Map> result = mapper.readValue(cmdOut.getOutput(), List.class);
+                    leader = result.stream().filter(row -> "LEADER".equals(row.get("role")))
+                        .map(row -> row.get("rpc-addresses"))
+                        .map(Objects::toString).findFirst()
+                        .orElseThrow(() -> new IllegalStateException("no leader master found"));
+                    return true;
+                });
+        }
+        return leader;
+    }
+
+    @Override
+    public boolean tableExists(String tableName) {
+        final PodShellOutput cmdOut = runCommand("table", "list", "-tables=" + tableName, getMasters());
+        handleOut(cmdOut);
+        return StringUtils.isNotBlank(cmdOut.getOutput()) && cmdOut.getOutput().contains(tableName);
+    }
+
+    @Override
+    public void createTable(Map<String, Object> tableDefinition) {
+        try {
+            final PodShellOutput cmdOut = runCommand("table", "create", getMasters(), mapper.writeValueAsString(tableDefinition));
+            handleOut(cmdOut);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public List<String> listRows(String tableName, List<String> columns, String predicates) {
+        final List<String> commands = new ArrayList<>(List.of("table", "scan", getMasters(), tableName));
+        if (columns != null && !columns.isEmpty()) {
+            commands.add("-columns=" + String.join(",", columns));
+        }
+        if (StringUtils.isNotBlank(predicates)) {
+            commands.add("-predicates=" + predicates);
+        }
+        final PodShellOutput cmdOut = runCommand(commands);
+        handleOut(cmdOut);
+        LOG.debug("table rows:\n{}", cmdOut.getOutput());
+        return cmdOut.getOutputAsList().stream().filter(line -> line.startsWith("("))
+            .map(line -> line.replaceAll("\\(", ""))
+            .map(line -> line.replaceAll("\\)", ""))
+            .toList();
+    }
+
+    @Override
+    public void close() {
+        leader = null;
+    }
+
+    /**
+     * Run command in POD
+     * @param commands String[], the commands without initial kudu
+     * @return @{@link PodShellOutput }
+     */
+    private PodShellOutput runCommand(final String... commands) {
+        final String[] com = new String[commands.length + 1];
+        com[0] = "kudu";
+        System.arraycopy(commands, 0, com, 1, com.length - 1);
+        return OpenshiftClient.get().podShell(OpenshiftClient.get().getPod(clientPodName)).execute(com);
+    }
+
+    private PodShellOutput runCommand(final List<String> commands) {
+        return runCommand(commands.toArray(new String[0]));
+    }
+
+    private static void handleOut(PodShellOutput cmdOut) {
+        if (StringUtils.isNotBlank(cmdOut.getError())) {
```

Comment:
why is this check not part of the runCommand method, since you then call `handleOut` after every `runCommand`? then you can return the output directly in the `runCommand` method

Reply:
done